otherlibs/stdune/src/io.ml

```diff
@@ -72,6 +72,131 @@ let copy_channels =
+              with
+              | fd_dst -> Ok (fd_dst, src_stat.st_size)
+              | exception exn -> (
+                Unix.close fd_src;
+                match exn with
+                | Unix.Unix_error (Unix.EISDIR, _, _) -> Error `Dst_is_a_dir
+                | _ -> Error (`Exn (Exn_with_backtrace.capture exn)))
+            in
+            (fd_src, fd_dst, src_size)))
+    in
+    fun ?chmod ~src ~dst () ->
+      (* All of this exception translation is done for now to maintain the
+         same error messages as the other file copying functions.
+
+         Eventually, we should stop using exceptions for signalling these
+         errors. But that's a bit of a large change since there's a lot of
+         exception catching to audit. *)
+      match setup_copy ?chmod ~src ~dst () with
+      | Error (`Exn exn) -> Exn_with_backtrace.reraise exn
+      | Error `Src_is_a_dir -> raise (Sys_error "Is a directory")
+      | Error `Dst_is_a_dir ->
+        let message = Printf.sprintf "%s: Is a directory" dst in
+        raise (Sys_error message)
+      | Error `Src_missing ->
+        let message = Printf.sprintf "%s: No such file or directory" src in
+        raise (Sys_error message)
+      | Ok (fd_src, fd_dst, src_size) ->
+        Exn.protectx (fd_src, fd_dst, src_size)
+          ~f:(fun (src, dst, src_size) -> sendfile ~src ~dst src_size)
+          ~finally:(fun (src, dst, _) ->
+            Unix.close src;
+            Unix.close dst)
+
+  let copyfile ?chmod ~src ~dst () =
+    let src_stats =
+      match Unix.stat src with
+      | exception Unix.Unix_error (Unix.ENOENT, _, _) ->
+        let message = Printf.sprintf "%s: No such file or directory" src in
+        raise (Sys_error message)
+      | { st_kind = S_DIR; _ } -> raise (Sys_error "Is a directory")
+      | stats -> stats
+    in
+    (try copyfile src dst with
+    | Unix.Unix_error (Unix.EPERM, "unlink", _) ->
+      let message = Printf.sprintf "%s: Is a directory" dst in
+      raise (Sys_error message)
+    | Unix.Unix_error (Unix.ENOENT, "realpath", _) ->
+      let message = Printf.sprintf "%s: No such file or directory" src in
+      raise (Sys_error message));
+    match chmod with
+    | None -> ()
+    | Some chmod -> src_stats.st_perm |> chmod |> Unix.chmod dst
+
+  let copy_file_portable ?chmod ~src ~dst () =
+    Exn.protectx (setup_copy ?chmod ~src ~dst ()) ~finally:close_both
+      ~f:(fun (ic, oc) -> copy_channels ic oc)
+
+  let copy_file_fast =
+    match available with
+    | `Sendfile -> sendfile
+    | `Copyfile -> copyfile
+    | `Nothing -> copy_file_portable
+
+  let copy_file_impl = ref copy_file_fast
+
+  let copy_file_portable ?chmod ~src ~dst () =
+    Exn.protectx (setup_copy ?chmod ~src ~dst ()) ~finally:close_both
+      ~f:(fun (ic, oc) -> copy_channels ic oc)
+
+  let copy_file ?chmod ~src ~dst () = !copy_file_impl ?chmod ~src ~dst ()
+end
+
+let set_copy_impl m =
+  Copyfile.copy_file_impl :=
```

Comment:
Not that it matters much but it seems faster to `match` on every call instead of following the pointer in the `ref`: the former benefits from branch prediction, so it's essentially free, while the latter always requires an indirection. Am I missing something?